### DISCUSSION TITLE
package `LinearAlgebra` not found

### DISCUSSION THREAD

**Discussion by ja72** [fonsp/Pluto.jl#2701] [Questions about using Pluto] [answered]
I am new to Pluto and very excited to use it. However, I hit a snag when I was unable to import the `LinearAlgebra` package into my worksheet.

![image](https://github.com/fonsp/Pluto.jl/assets/22509289/710db096-5fa7-4947-a11f-6f60cfa4b9eb)

In the underlying Julia installation of course I have [`LinearAlgebra`](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/) installed as a package and it works as intended.

So what I am missing here with Pluto?

**Comment by holomorphism** [comment]
What version of Pluto are you using?
I don't know what is causing the problem either, but I was wondering why you are not seeing the error message above the cell.
I believe (the latest version of) Pluto should display the following error message above the cell if it fails to load the package.

![1](https://github.com/fonsp/Pluto.jl/assets/3076585/b9123a8b-65fb-4db2-a945-33b8e5cc045c)

The fact that it is not showing it makes me think that what appears to be an error is just a display error and that LinearAlgebra is actually loading correctly.

For example, what do you see when you run the following code in the following cell? (`dot` is a function in `LinearAlgebra`)

```julia
dot([1,2],[3,4])
```

If the output is correct and not an error, `LinearAlgebra` should be loaded.

**Reply by pankgeorg** [reply]
What version of Julia are you using? Standard libraries didn't register as dependencies up until recently but now they do.

**Reply by ja72** [reply]
Γιά-σου Παναγιωτη (Hello @pankgeorg). 

I have `Pluto v0.17.7` installed within Julia 1.9.3

when I ran `status` in the package environment, I noticed that `LinearAlgebra` wasn't listed.

```
(@v1.9) pkg> status
Status `C:\Users\jalex\.julia\environments\v1.9\Project.toml`
⌅ [46ada45e] Agents v4.3.1
  [7820620d] AgentsPlots v0.4.2
  [ffbed154] DocStringExtensions v0.9.3
  [7073ff75] IJulia v1.24.2
  [91a5bcdd] Plots v1.39.0
⌃ [c3e4b0f8] Pluto v0.17.7
  [0756cd96] Quaternionic v1.4.0
  [90137ffa] StaticArrays v1.6.5
⌃ [0c5d862f] Symbolics v5.5.1
Info Packages marked with ⌃ and ⌅ have new versions available, but those with ⌅ are restricted by compatibility constraints from upgrading. To see why use `status --outdated`
```

after running the command

```
(@v1.9) pkg> add LinearAlgebra
   Resolving package versions...
    Updating `C:\Users\jalex\.julia\environments\v1.9\Project.toml`
  [37e2e46d] + LinearAlgebra
  No Changes to `C:\Users\jalex\.julia\environments\v1.9\Manifest.toml`
```

it is now listed, and so I am trying Pluto again.

The error is still there since the following cell in Pluto 

```
let
	using LinearAlgebra;
	using StaticArrays;
	using Quaternionic;
end
```

produced the following output in the REPL

```
Press Ctrl+C in this terminal to stop Pluto

   Resolving package versions...
    Updating `C:\Users\jalex\AppData\Local\Temp\jl_ebfLPl\Project.toml`
  [0756cd96] + Quaternionic v1.4.0
  [90137ffa] + StaticArrays v1.6.5
    Updating `C:\Users\jalex\AppData\Local\Temp\jl_ebfLPl\Manifest.toml`
  [59287772] + Formatting v0.4.2
  [b964fa9f] + LaTeXStrings v1.3.1
  [23fbe1c1] + Latexify v0.16.1
  [1914dd2f] + MacroTools v0.5.11
  [bac558e1] + OrderedCollections v1.6.2
  [aea7be01] + PrecompileTools v1.2.0
  [21216c6a] + Preferences v1.4.1
  [0756cd96] + Quaternionic v1.4.0
  [ae029012] + Requires v1.3.0
  [90137ffa] + StaticArrays v1.6.5
  [1e83bf80] + StaticArraysCore v1.4.2
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [b77e0a4c] + InteractiveUtils
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [d6f4376e] + Markdown
  [de0858da] + Printf
  [9a3f8284] + Random
  [ea8e919c] + SHA v0.7.0
  [9e88b42a] + Serialization
  [fa267f1f] + TOML v1.0.3
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [e66e0078] + CompilerSupportLibraries_jll v1.0.5+0
  [4536629a] + OpenBLAS_jll v0.3.21+4
  [8e850b90] + libblastrampoline_jll v5.8.0+0
┌ Warning: The project dependencies or compat requirements have changed since the manifest was last resolved.
│ It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
└ @ Pkg.API C:\Users\jalex\AppData\Local\Programs\Julia-1.9.3\share\julia\stdlib\v1.9\Pkg\src\API.jl:1687
```

and the familiar error that I originally posted about.

**Reply by ja72** [reply]
@holomorphism - it is not just a display error.

![image](https://github.com/fonsp/Pluto.jl/assets/22509289/5729b24d-c1a3-4296-9d12-910a57f2acba)

**Reply by pankgeorg** [reply]
Γεια! Can you `]up Pluto` in the REPL? And try to add `import Pkg;Pkg.add("LinearAlgebra")` somewhere, it will probably also help!

**Reply by ja72** [reply]
After manually adding `LinearALgebra` in the REPL (even though I have it in my `startup.jl` already) and running Pluto the problem is now visual only.

![image](https://github.com/fonsp/Pluto.jl/assets/22509289/f14e6423-8d7e-4c55-9273-4eaa1b265dce)

**Comment by ja72** [comment]
It turns out @pankgeorg was correct. `Pluto` did not load the correct version somehow.

What I did to fix the issue was

 1. Delete the `.julia` folder (or rename to `.julia (old)`).
 2. Run `Julia REPL` to recreate the envirnoments
 3. Add needed packages from package manager mode `]add LinearAlgebra`
 4. Add Pluto `add Pluto`
 5. Run `status` to check everything
 6. Close the REPL
 7. (Optional) Edit the `.julia/config/startup.jl` file to load Pluto in `REPL` every time

    ```
	import Pkg
	let
		pkgs = ["LinearAlgebra", "DocStringExtensions", "StaticArrays", "Symbolics", "Pluto"]
		for pkg in pkgs
			if Base.find_package(pkg) === nothing
				Pkg.add(pkg)
			end
		end
		using LinearAlgebra
		using DocStringExtensions
		using StaticArrays
		using Pluto
		show(pkgs)
		println(" Loaded")
	end
    ```
       
 8. Start the REPL and launch Pluto

    ```
    julia> Pluto.run()
    ```

The Pluto environment has not the updated look

![image](https://github.com/fonsp/Pluto.jl/assets/22509289/8903092f-3737-4fd9-a19a-ea0f0b95f968)

and the built-in libraries work as intended

```
let
	using LinearAlgebra;
	using StaticArrays;
	using Quaternionic;
end
```

with test code `dot([1,2],[3,4])` producing `11`.

**Reply by fonsp** [reply]
@pankgeorg What do you think? I don't see how the edits to startup.jl solved it because Pluto doesn't use startup.jl nor the global env. Deleting `.julia` might have been the solution...

**Reply by fonsp** [reply]
Ah it's an old Pluto version.

**Reply by pankgeorg** [reply]
I think it was a very old Pluto version, and probably some packages are being held back by others in the global env

**Comment by fonsp** [comment, marked as answer]
I'm not able to reproduce this on Julia 1.9.3 MacOS with latest Pluto, and it looks like @holomorphism also didn't reproduce on Windows in https://github.com/fonsp/Pluto.jl/discussions/2701#discussioncomment-7543461 .

@ja72 it looks like you are using Pluto 0.17.7, which is almost 2 years old, and released before Julia 1.9, so it might not be fully supported on Julia 1.9. Are you not able to update Pluto to the latest version?

**Reply by ja72** [reply]
You are correct.

The problem is fixed when I start with a fresh new Julia environment and a new Pluto package add.
